### ISSUE TITLE
Ninja Baseball Batman "Fix" IRQ/Raster Issues?

### ISSUE THREAD

**Issue by KMFDManic** [libretro/mame2003-plus-libretro#1874]
@mahoneyt944 @grant2258 @arcadez2003 

I am not entirely sure we addressed this before.  But, it seems like it should be relatively reasonably feasible to address the slowdown issues present in Ninja Baseball Batman, which appear to be IRQ/Raster Effect related, of which quite a few fixes went into MAME for.  Hopefully, the entire driver wouldn't need updated to take advantage of some of these fixes.  I always assumed it was just related to protection issues.  But, now, seems more so it was slowdown for Raster Effects, and a few other pesky issues, which may be very well possible to fix-up in the 2003 code database...well, hopefully!  One can never have enough IREM awesomeness:)  Sorry if this has been brought up before.  I didn't see much related to it for 2003, in my quick search.

M92 Driver, for Ninja Baseball, which I am certain you all already know:)  FBNEO has quite a few fixes, as does MAME.

Bugs:

- nbbatman, bbatmanu: Effects in Ninja Baseball Batman flicker and display incorrectly. samsho2 (ID 06021)

- Clone nbbatman2bl: AddressSanitizer: heap-buffer-overflow with -aviwrite. Firewave (ID 05865)

WIP:

- 0.148u5: Smitdogg replaced bad gfx dumps in clone Ninja Baseball Bat Man II (bootleg).

- 0.147u1: hap changed description of clone '(bootleg/hack)' to 'Ninja Baseball Bat Man II (bootleg)'.

- 0.146u5: Smitdogg, The Dumping Union and David Haywood added clone Ninja Baseball Bat Man II (bootleg/hack).

- 0.146u4: Arzeno Fabrice added Ninja Baseball Bat Man (World). Renamed (nbbatman) to (nbbatmanu).

- 28th January 2012: Smitdogg - We got a booty of Ninja Baseball Bat Man called Ninja Baseball Bat Man II. Changes: Title screen changed. Overall attack speed might be slightly faster? Button C added, it changes the player's size and attack style if you tap it, sometimes if you hold it it does a weird combo attack where you change around sizes while fighting. Sound seems to be changed to samples and it's a shorter music cut just repeated over and over. Maybe other changes.  Not sure which resource you find generally best to try to verify some of the additions or fix-ups!

https://github.com/libretro/FBNeo/commits/master/src/burn/drv/irem/d_m92.cpp?after=a8f500c8819bf33cbdb3f038ac1f981290bcfee4+34

- 0.143u7: Mamesick fixed sounds freezes or halts after a time in Ninja Baseball Bat Man.

- 0.142u3: Changed description to 'Ninja Baseball Bat Man (US)'. Fixed maincpu rom loading.

-  8th June 2009: Guru - Ninja Baseball Batman (1993 Irem) arrived. Thanks to Gp-Lee.

- 0.128u6: Couriersud fixed raster effects in Ninja Baseball Batman by altering the V33 prefetch timing (just a guess as no documentation has been found).

- 0.105u4: Twisty added 'Difficulty' dipswitch.

- 0.99u7: Pierpaolo Prazzoli added missing opcode 0 in leagueman_decryption_table, used in Ninja Baseball Batman (machine\irem_cpu.c).

- 0.56: Added player and coin slots 3 and 4. Added dipswitches 'Lives', 'Any Button to Start' and 'Demo Sounds'.

- 21st September 2001: Bryan McPhail added 3rd and 4th player controls to Ninja Baseball Batman.

- 0.55: Bryan McPhail added Ninja Baseball Batman (US) (Irem 1993) and clone Yakyuu Kakutou League-Man (Japan).

- 26th August 2001: Bryan McPhail improved the IREM M-92 driver, making Ninja Baseball Batman work.

- 0.36b6: Bryan McPhail added Ninja Baseball Batman (Testdriver) and clone Yakyuu Kakutou League-Man (Japan). Very picky about interrupts. Doesn't work!

- 24th September 1999: Oliver Bergmann sent in a NEC V20, V30, V33 cpu core, and Bryan McPhail submitted an Irem M92 driver with support for Blade Master, Gunforce, Lethal Thunder, Hook, Undercover Cops, R-type Leo, Major Title 2 and In The Hunt, but none of the games have sound because of an encrypted sound CPU.

Mike Tyson and Jake Paul fight next week on Netflix...Gonna be quite an interesting one to see:)

**Comment by grant2258** [comment]
I think this only happens at the map screens well I remember it that way in my head.

**Comment by KMFDManic** [comment]
@grant2258 Correct, which appears to be due to Raster Effects at those points.  I believe the tilesets are also offset, amongst a few other minor issues.  On Lower Spec Platforms, the slowdown at the map junction points can be quite irritating, with how long the game lags for.  The fixes appear to be in place for FBNEO, as the Map Loading portions are seamless...But, of course, FBNEO isn't as optimal on Lower Spec for M92 games, unfortunately.  Seems like, if we don't actually have to update the driver, itself, could be trivial to just address these hopefully minor to fix conflicts.  This could, of course, parlay into fix-ups for a few other games on the same driver.

Hope all is well, by the way!

**Comment by grant2258** [comment]
Yes all is well am in the middle of possibly changing jobs managed to get a job opportunity in the council. Ill officially be employed by the government if everything works out. We all just need to go where the money is these days, ive not completely made my mind up ive passed the usual background checks ect  and its multi stages interviews. If it turns out a better deal ill take it.

**Comment by arcadez2003** [comment]
This is the on;y mention of a fix for the raster effects in the MAME WiP when it comes to the game.......
https://github.com/arcadez2003/mame/commit/4ffa2901ceed2884ddbb9c5046712427ba3b871f

From memory i looked into adding the fix here but our NEC CPU is totally different so it was hard
to pindown for me atleast where in the code to splice it in maybe FBN has an easier to implement
fix for this.??

**Comment by KMFDManic** [comment]
@arcadez2003 Looking at the fixes in FBNEO, some seem to be related to kludge and NCycles and Vez IRQ?  Maybe @dinkc64 could shed some light?

**Comment by arcadez2003** [comment]
Aye he's the man to ask alright so why dont ya :)

**Comment by dinkc64** [comment]
Heyo arcadez, sorry buddy, I missed ya on ye olde chatter earlier.  Damned time change (DST) is messing with me...

Regarding nbbatman: setting up the 8529 pic (programmable interrupt chip) fixed all the problems with nbbatman and got rid of all the kludges regarding messing about with cycles.  I recommend doing that, and totally ignore the old kludge-fixes, which were kinda horrible way to do things.

best regards,
- dink

**Comment by dinkc64** [comment]
Addendum:
the pic8529 fixes the map screen issue
a slight overclocking (9.96mhz) will help with the raster effects

**Comment by arcadez2003** [comment]
That's for the advice @dinkc64 i'll pop on the chat tomorrow and BTW i came on the say hello not
to bug ya for a fix here as i felt it was @KMFDManic who should ask since he dropped your name
into this conversation :)

**Comment by KMFDManic** [comment]
@dinkc64 Much appreciated, your feedback.  Was a pleasure collaborating with ya in the past on Coleco Stuff:)  I have been keeping up on all the commits you've done on FBNEO, and Arcade Perfection comes with both FBNEO/2003 Plus Combos, to pretty much cover all the bases:)  No need for "Current" MAME, haha:)

**Comment by dinkc64** [comment]
@KMDFManic, ahh yes, that was a good time :)  I still have a great time playing these old games, but, personally I prefer the latest MAME - it's debugger can't be beat, and it remembers the windows (for debugging).

best regards,
- dink

**Comment by KMFDManic** [comment]
@mahoneyt944 @arcadez2003 @grant2258 I got back to Ninja Baseball Batman, analyzing what dink had done on FBNEO, and implemented the coding fixes into Xtreme.  Did a test back to back with Plus/Xtreme, as well as OLD Xtreme and Xtreme with the raster fixes...The preload of stages is now instantaneous and slowdown is gone!  I will commit to my repo, so you can see the fix and try it for Plus, see if it suits ya.  But, as far as I am concerned, the slowdown being gone is awesome..It now runs closer to how it should run!  Thanks, again, @dinkc64 your your info on this.  Good to go, with maybe a slight touch up, if at all necessary, once the Plus guys check things out:)

**Comment by KMFDManic** [comment]
@arcadez2003 @mahoneyt944 @grant2258 

https://github.com/KMFDManic/mame2003-xtreme/commit/12514eee27beecf72db53ca59fc58a54db12fdda

Test wise, slowdown is gone on preload for both Plus/Xtreme.  There is a very minor flicker in the transition, which is likely due to the slightly offset timing between IRQ raster and vblank interrupts.  Since it won't hurt things any, may add a configurable Core Option specifically for the m92 driver, that can adjust the OC rate, specifically for that driver, to more quickly find the most optimal default.

But, as of now, going to Plus/OLD Xtreme, then testing this implementation, the preload of stages is WAY WAY faster.  Try for yourself:)  Feel free to let me know if any adjustments should be made.  

Not sure why I sat on this one for so long, probably because I got back into the Custom OST stuff for a bit.  But, wanted to at least get this one fixed up, and things seem to be considerably better on my end!

I will probably get a video demonstration up by tomorrow, since it is quite a significant difference in playability with the much faster raster loading!  Like grant2258 had said before, kludge fixes are like duct tape fixes...they are what they are.  I'd personally rather run this game more seamlessly without the level preload slowing down for 5+ seconds on my lower spec!!!  No issues with that anymore:)

**Comment by grant2258** [comment]
im very surprised it wasn't the NEC CPU updates, well ill be a monkeys uncle not the first time ive been wrong haha.

**Comment by grant2258** [comment]
any reason this isint being added been a very long time with no feedback

**Comment by arcadez2003** [comment]
> any reason this isint being added been a very long time with no feedback?

Well since no one did a pr we must have forgotten about it :)
